src/LanguageServer/Protocol/Workspaces/LspMiscellaneousFilesWorkspace.cs

```diff
@@ -58,13 +74,44 @@ internal sealed class LspMiscellaneousFilesWorkspace(ILspServices lspServices, I
         }
 
         var sourceTextLoader = new SourceTextLoader(documentText, documentFilePath);
+        var services = Services.SolutionServices;
+        var languageServices = services.GetLanguageServices(languageInformation.LanguageName);
 
-        var projectInfo = MiscellaneousFileUtilities.CreateMiscellaneousProjectInfoForDocument(
-            this, documentFilePath, sourceTextLoader, languageInformation, documentText.ChecksumAlgorithm, Services.SolutionServices, []);
+        var parseOptions = languageServices.GetService<ISyntaxTreeFactoryService>()?.GetDefaultParseOptionsWithLatestLanguageVersion();
+        var isFileBasedProgram = checkIsFileBasedProgram();
+        if (parseOptions != null && isFileBasedProgram)
+        {
+            parseOptions = parseOptions.WithFeatures([new("FileBasedProgram", "true"), .. parseOptions.Features]);
+            // TODO2: maybe initialize analyzers here?
+        }
+
+        var projectInfo = MiscellaneousFileUtilities.CreateMiscellaneousProjectInfoForDocumentCore(
+            this,
+            documentFilePath,
+            sourceTextLoader,
+            languageInformation,
+            documentText.ChecksumAlgorithm,
+            services,
+            languageServices,
+            parseOptions,
+            // TODO2: does language server know where the SDK is?
+            metadataReferences: [.. Net90.References.All]);
         OnProjectAdded(projectInfo);
 
         var id = projectInfo.Documents.Single().Id;
         return CurrentSolution.GetRequiredDocument(id);
+
+        bool checkIsFileBasedProgram()
+        {
+            var languageServices = Services.SolutionServices.GetLanguageServices(languageInformation.LanguageName);
+            if (parseOptions is not CSharpParseOptions csharpOptions)
+                return false;
+
+            var tree = CSharpSyntaxTree.ParseText(text: documentText, options: csharpOptions, path: documentFilePath);
+            var root = tree.GetRoot();
+            var descendants = root.DescendantNodes(descendIntoChildren: node => node.IsKind(SyntaxKind.CompilationUnit));
+            return descendants.OfType<GlobalStatementSyntax>().Any() || descendants.Where(d => d.IsKind(SyntaxKind.ShebangDirectiveTrivia)).Any();
```

Comment:
this doesn't match the logic in another file.

Reply:
code deleted.